core/services/workflows/engine.go

```diff
@@ -506,73 +546,42 @@ func (e *Engine) handleStepUpdate(ctx context.Context, stepUpdate store.Workflow
 		return err
 	}
 	l := e.logger.With(eIDKey, state.ExecutionID, sRKey, stepUpdate.Ref)
+	workflowFullyIsProcessed, status, err := e.isWorkflowFullyProcessed(state)
+	if err != nil {
+		return err
+	}
 
-	switch stepUpdate.Status {
-	case store.StatusCompleted:
-		stepDependents, err := e.workflow.dependents(stepUpdate.Ref)
-		if err != nil {
-			return err
-		}
-
-		// There are no steps left to process in the current path, so let's check if
-		// we've completed the workflow.
-		if len(stepDependents) == 0 {
-			workflowCompleted := true
-			err := e.workflow.walkDo(workflows.KeywordTrigger, func(s *step) error {
-				step, ok := state.Steps[s.Ref]
-				// The step is missing from the state,
-				// which means it hasn't been processed yet.
-				// Let's mark `workflowCompleted` = false, and
-				// continue.
-				if !ok {
-					workflowCompleted = false
-					return nil
-				}
-
-				switch step.Status {
-				case store.StatusCompleted, store.StatusErrored, store.StatusCompletedEarlyExit:
-				default:
-					workflowCompleted = false
-				}
-				return nil
-			})
-			if err != nil {
-				return err
-			}
-
-			if workflowCompleted {
-				return e.finishExecution(ctx, state.ExecutionID, store.StatusCompleted)
-			}
+	if workflowFullyIsProcessed {
+		switch status {
+		case store.StatusCompleted:
+			l.Info("workflow finished")
+		case store.StatusErrored:
+			l.Info("execution errored")
+		case store.StatusCompletedEarlyExit:
+			l.Info("execution terminated early")
+			// NOTE: even though this marks the workflow as completed, any branches of the DAG
+			// that don't depend on the step that signaled for an early exit will still complete.
+			// This is to ensure that any side effects are executed consistently, since otherwise
+			// the async nature of the workflow engine would provide no guarantees.
 		}
+		return e.finishExecution(ctx, state.ExecutionID, status)
+	}
 
-		// We haven't completed the workflow, but should we continue?
-		// If we've been executing for too long, let's time the workflow out and stop here.
-		if state.CreatedAt != nil && e.clock.Since(*state.CreatedAt) > e.maxExecutionDuration {
-			l.Info("execution timed out")
-			return e.finishExecution(ctx, state.ExecutionID, store.StatusTimeout)
-		}
+	// We haven't completed the workflow, but should we continue?
+	// If we've been executing for too long, let's time the workflow out and stop here.
+	if state.CreatedAt != nil && e.clock.Since(*state.CreatedAt) > e.maxExecutionDuration {
```

Comment:
This should be handled alongside the workflowIsFullyProcessed check, using the same technique as errored: if we've timedout, let's propagate the timeout status forward, and finish if we've processed everything.

If we haven't then something else is still executing and we can rely on them to finish the workflow.